pydal/adapters/base.py

```diff
@@ -618,49 +624,60 @@ def _select_wcols(self, query, fields, left=False, join=False,
         #: build joins (inner, left outer) and table names
         if join:
             (
+                # FIXME? ijoin_tables is never used
                 ijoin_tables, ijoin_on, itables_to_merge, ijoin_on_tables,
-                iimportant_tablenames, iexcluded
-            ) = self._build_joins_for_select(tablenames, join)
+                iimportant_tablenames, iexcluded, itablemap
+            ) = self._build_joins_for_select(tablemap, join)
+            # FIXME: check for name conflicts
+            tablemap.update(itablemap)
         if left:
             (
                 join_tables, join_on, tables_to_merge, join_on_tables,
-                important_tablenames, excluded
-            ) = self._build_joins_for_select(tablenames, left)
+                important_tablenames, excluded, jtablemap
+            ) = self._build_joins_for_select(tablemap, left)
+            # FIXME: check for name conflicts
+            tablemap.update(jtablemap)
+        current_scope = outer_scoped + list(tablemap)
+        table_alias = lambda name: tablemap[name].query_name(current_scope)[0]
         if join and not left:
             cross_joins = iexcluded + list(itables_to_merge)
-            sql_t = '%s' % self.table_alias(cross_joins[0])
-            for t in cross_joins[1:]:
-                sql_t += ' %s' % self.dialect.cross_join(self.table_alias(t))
-            for t in ijoin_on:
-                sql_t += ' %s' % self.dialect.join(t)
+            tokens = [table_alias(cross_joins[0])]
+            tokens += [self.dialect.cross_join(table_alias(t))
+                    for t in cross_joins[1:]]
+            # FIXME: properly expand subqueries in the ON() expressions
+            tokens += [self.dialect.join(t) for t in ijoin_on]
+            sql_t = ' '.join(tokens)
         elif not join and left:
             cross_joins = excluded + list(tables_to_merge)
-            sql_t = '%s' % self.table_alias(cross_joins[0])
-            for t in cross_joins[1:]:
-                sql_t += ' %s' % self.dialect.cross_join(self.table_alias(t))
+            tokens = [table_alias(cross_joins[0])]
+            tokens += [self.dialect.cross_join(table_alias(t))
+                    for t in cross_joins[1:]]
+            # FIXME: WTF? This is not correct syntax at least on PostgreSQL
             if join_tables:
-                sql_t += ' %s' % \
-                    self.dialect.left_join(','.join([t for t in join_tables]))
-            for t in join_on:
-                sql_t += ' %s' % self.dialect.left_join(t)
+                tokens.append(self.dialect.left_join(','.join([table_alias(t)
+                        for t in join_tables])))
```

Comment:
Query syntax like `LEFT JOIN tab1,tab2` without any `ON` clause doesn't work in PostgreSQL and possibly other databases. If some databases do support this syntax, it should be moved to corresponding dialects. Don't forget to update the other copy of this code below as well.